KOGITO-2664: [DMN Designer] Multiple DRDs support

karreiro opened the pull request:
**JIRA**: [KOGITO-2664](https://issues.redhat.com/browse/KOGITO-2664): [DMN Designer] Multiple DRDs support

**Part of an ensemble**:
- https://github.com/kiegroup/kie-wb-common/pull/3379
- https://github.com/kiegroup/drools-wb/pull/1416
- https://github.com/kiegroup/kie-wb-distributions/pull/1054

## Artifacts
- **Business Central**: [WAR file v6](https://www.dropbox.com/s/6lovcz78ixv7a55/business-central-v6.war?dl=0)
- **VS Code**: [plugin v6](https://www.dropbox.com/s/ukfin8xsp5rl5sj/plugin-v6.vsix?dl=0)


## Demos

This PR covers the following JIRAs:

- [KOGITO-2665](https://issues.redhat.com/browse/KOGITO-2665): Marshaller - Adopt the first DMNDiagram as the DRG (author: @karreiro)
<img width="50%" src="https://user-images.githubusercontent.com/1079279/92027801-69abb180-ed39-11ea-83fa-ff09db191487.gif">
<img width="50%" src="https://user-images.githubusercontent.com/1079279/92027810-6ca6a200-ed39-11ea-91c8-1028635c722d.gif">

- [KOGITO-2965](https://issues.redhat.com/browse/KOGITO-2965): Graph - Palette changes (author: @danielzhe)
- [KOGITO-2666](https://issues.redhat.com/browse/KOGITO-2666): Graph - Create mechanism to switch context between graphs (author: @karreiro)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92027837-74fedd00-ed39-11ea-895c-02bd274b37f7.gif">

- [KOGITO-2669](https://issues.redhat.com/browse/KOGITO-2669): Decision Navigator - Split current navigator between the DRG and the multiple DRDs (author: @karreiro)
- [KOGITO-2670](https://issues.redhat.com/browse/KOGITO-2670): Decision Navigator - The DRD name must be editable into the decision navigator (author: @karreiro)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92027965-a5df1200-ed39-11ea-96a5-47709ed7d5ae.gif">

- [KOGITO-2991](https://issues.redhat.com/browse/KOGITO-2991): Graph - Keep DRG elements instances synchronised (author: @danielzhe)
- [KOGITO-2746](https://issues.redhat.com/browse/KOGITO-2746): Marshaller - Many nodes instances for a single DRG element (author: @karreiro)
- [KOGITO-2671](https://issues.redhat.com/browse/KOGITO-2671): Decision Components - The decision components widget must show DRGs from the current model (author: @danielzhe)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92028022-bdb69600-ed39-11ea-9cc1-eac328b9d185.gif">

- [KOGITO-2672](https://issues.redhat.com/browse/KOGITO-2672): Graph - Node remove command (trash icon) should not always delete the DRG element (author: @danielzhe)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92028034-c313e080-ed39-11ea-8ce1-eed374fd1c2c.gif">

- [KOGITO-2674](https://issues.redhat.com/browse/KOGITO-2674): Context menu component (@vpellegrino)
- [KOGITO-2675](https://issues.redhat.com/browse/KOGITO-2675): Context menu - Show the DRD clickable icon when node is selected (@vpellegrino)
- [KOGITO-3022](https://issues.redhat.com/browse/KOGITO-3022): Context menu - When multiple nodes are selected, showing context menu with right click (@vpellegrino)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92028054-cad38500-ed39-11ea-8224-3f0403c0d89a.gif">

- [KOGITO-2676](https://issues.redhat.com/browse/KOGITO-2676): Graph - The back link must show the DRD name (@vpellegrino)
- [KOGITO-2668](https://issues.redhat.com/browse/KOGITO-2668): Graph - When a DRD is focused, the name must appear above the canvas (@vpellegrino)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92028064-d030cf80-ed39-11ea-8829-9ebf343eda29.gif">

- [KOGITO-2964](https://issues.redhat.com/browse/KOGITO-2964): Marshaller - Use a single Marshaller on all DMN webapps (@karreiro)

<img width="50%" src="https://user-images.githubusercontent.com/1079279/92028789-facf5800-ed3a-11ea-9e64-3dc63ca1bdcc.png">

---

<pre>
To retest a PR or trigger a specific build please add a comment:

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</pre>

karreiro commented:
@danielzhe @vpellegrino Now this branch is up to date with master (I also included your commit here, Daniel) :-)

karreiro commented:
Hi @danielzhe @jomarko @vpellegrino,

I just updated this PR merging our DRD-related branches, I our code is ready for review.

Tomorrow, I expect to convert this draft into a regular PR, build the artifacts, and leave their URL here (for both, BC and Kogito). Then we all can test it, find bugs, and improve it.

I've tried to divide our work into 24 meaningful commits to support the review. The largest commits are the first ones because we're moving the marshaller from Kogito to `kie-wb-common-dmn-client`. Thus, I recommend we review this PR by analyzing commits chronologically :-)

I don't expect we will have structural changes in the code; just minor fixes applied on _new_ commits (instead of squashing).

Thank you.

jstastny-cz commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNContentServiceImpl.java:
```diff
@@ -0,0 +1,176 @@
+                                            overviewLoader.loadOverview(path));
+    }
+
+    private ProjectMetadata buildProjectMetadataInstance(final Path path,
```
is there an expectation of reusing the method? If not it IMO adds unnecessary lines of code. But no big deal if you prefer this way.

karreiro replied:
I agree. Fixed! Thanks.

jstastny-cz reviewed (commented):
I've finished the first code review round now, will proceed with testing.

jstastny-cz commented:
jenkins do fdb

jstastny-cz commented:
@karreiro there seems to be misconfig in the 
```

karreiro commented:
Jenkins do fdb.

jstastny-cz commented:
Issue#1
- Overlapping Breadcrumbs
  1. Create DRD
  2. Navigate there
  3. Add DMN Decision
  4. Edit Decision - the breadcrumbs items are overlapping as per the screenshot
![Screenshot from 2020-09-09 16-26-08](https://user-images.githubusercontent.com/11939909/92611644-493fa200-f2b9-11ea-8181-d7753b587fac.png)

Issue#2
- Can't create DRD before an element is added to the diagram, or is not intuitive (couldn't find a way)

Issue#3
- The icon for DRD-related actions is not intuitive at all - it's the sign used commonly for "Share"
![Screenshot from 2020-09-09 16-36-09](https://user-images.githubusercontent.com/11939909/92612875-97a17080-f2ba-11ea-8f37-948b7d2ecd88.png)


Issue#4
- The Data Types are scoped on where it has been defined. How to share Data Type across DRDs or between DRD and DRG?

Issue#5
- The Decision Components are not easily distinguished  between DRDs or DRG, now it's easy to create nodes with same names - the default naming Decision-x is DRD-scoped, not global, thus I can end up easily with many such nodes. Then in Decision Components I can't really tell which is which (because the id in Decision Components list actually differ from id that user can see in the node details - not sure if expected)
![Screenshot from 2020-09-09 16-43-54](https://user-images.githubusercontent.com/11939909/92614116-f9aea580-f2bb-11ea-823b-65e1c82f79fd.png)

karreiro commented:
Jenkins do fdb.

karreiro commented:
Jenkins do fdb.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/drds/DMNDiagramTuple.java:
```diff
@@ -0,0 +1,41 @@
+        return stunnerDiagram;
+    }
+
+    public DMNDiagramElement getDMDNDiagram() {
```
Well spotted, thanks! Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/drds/DMNDiagramsSession.java:
```diff
@@ -0,0 +1,195 @@
+        return getSessionState().getDRGDiagram();
+    }
+
+    public DMNDiagramElement getDRGDMNDiagramElement() {
```
Yeah, I generally prefer to follow the convention of `getDmnDiagrams`, instead of `getDMNDiagrams`, it's more readable indeed. But, on `kie-wb-common-dmn`, the current convention is `getDMNDiagrams`.

We can change it in the future, but for this PR I just named the method from `getDRGDMNDiagramElement` to `getDRGDiagramElement `. I believe it improves the readability a little bit.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/tree/DecisionNavigatorTreeView.java:
```diff
@@ -247,11 +205,45 @@ public void onIconClick(final ClickEvent event) {
+        }
+
+        @EventHandler("input-text")
+        public void onInputTextBluer(final BlurEvent event) {
```
Fixed! Thanks.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/preview/PreviewDiagramScreen.java:
```diff
@@ -132,18 +130,15 @@ void closePreview() {
     void onCanvasSessionOpened(final @Observes SessionOpenedEvent sessionOpenedEvent) {
-        checkNotNull("sessionOpenedEvent", sessionOpenedEvent);
-        showPreview(sessionOpenedEvent.getSession());
+        // TODO {karreiro}: Fix mini diagram preview.
```
Fixed! Thanks.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/drd/DRDContextMenuService.java:
```diff
@@ -0,0 +1,152 @@
+
+    public void addToNewDRD(final Collection<Node<? extends Definition<?>, Edge>> selectedNodes) {
+
+        // TODO: https://issues.redhat.com/browse/KOGITO-2992
```
Fixed! Thanks.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/preview/PreviewDiagramScreenTest.java:
```diff
@@ -95,6 +96,7 @@ public void testOnClose() {
     }
 
     @Test
+    @Ignore("TODO {karreiro}: Fix mini diagram preview.")
```
Fixed! Thanks.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorPresenterTest.java:
```diff
@@ -204,202 +172,29 @@ public void testGetTreePresenter() {
         doReturn(items).when(presenter).getItems();
 
+        presenter.disableRefreshHandlers();
+        presenter.refreshTreeView();
```
Because, we're disabling invoking, then enabling and invoking. Thus, in the end of the test we're verifying that the `setupItems(items)` is called once (since the first invocation was disabled).

karreiro commented:
Thank you for the review, @jstastny-cz.

Regarding the issues you've mentioned:

- **Issue#1**: Fixed! Thanks for reporting!

- **Issue#2**/**Issue#3**: Both comments have interesting points. At this first version, we're trying to achieve something as similar as possible to the prototypes provided by the UX team. I'm going to hold these feedbacks since we will have others about DRDs, and we can handle all of them on a future release. For the scope of this PR, I believe we can keep this component as they are.

- **Issue#4**: I'm taking a look at this to incorporate a fix on this PR. Data Types are global for the model; they shouldn't be scoped by each DRD.

- **Issue#5**: Good catch. Ideally, when users have a "Decision-1" on a DRD-1, and they create a new decision in the DRD-2, we would expect this new decision to be called "Decision-2". However, today the counter is being "reseted" for each DRD. I've raised this JIRA https://issues.redhat.com/browse/KOGITO-3306, they we can handle it in another PR.

Wrt the log you've linked (thanks btw), the issue was fixed as well (we were including a JS file that shouldn't be included on non-development web apps).

@jstastny-cz, there are still some comments pending to answer. I will catch up them by tomorrow (I tried to address the difficult ones first hehe) :)

karreiro commented:
Jenkins do fdb.

jstastny-cz commented:
Issue#6
1. in BC open Traffic Violation Sample
2. click Fine Drd actions -> Create
3. rename DRD in Decision Navigator
4. Unexpected error appears with following log in server.log:
  `
2020-09-10 10:57:15,320 ERROR [org.kie.workbench.common.services.backend.logger.GenericErrorLoggerServiceImpl] (default task-11) Error from user: testadmin Error ID: -163054630 Location: LibraryPerspective|$ProjectScreen,org.kie.workbench.common.screens.messageconsole.MessageConsole[Worg.kie.dmn.decision.navigator,],DMNDiagramEditor?path_uri=default://master@MySpace/Traffic_Violation/src/main/resources/org/kie/example/traffic/traffic_violation/Traffic%2520Violation.dmn&file_name=Traffic%20Violation.dmn&has_version_support=true Exception: Uncaught exception: Exception caught: CDI Event exception: CommandType=CDIEvent, BeanType=org.kie.workbench.common.dmn.client.docks.navigator.drds.DMNDiagramSelected, BeanReference=org.kie.workbench.common.dmn.client.docks.navigator.drds.DMNDiagramSelected@667a, FromClient=1 sent to [unavailable] Caused by: CDI Event exception: CommandType=CDIEvent, BeanType=org.kie.workbench.common.dmn.client.docks.navigator.drds.DMNDiagramSelected, BeanReference=org.kie.workbench.common.dmn.client.docks.navigator.drds.DMNDiagramSelected@667a, FromClient=1 sent to [unavailable] Caused by: (NotFoundError) : Failed to execute 'removeChild' on 'Node': The node to be removed is no longer a child of this node. Perhaps it was moved in a 'blur' event handler?`
![Screenshot from 2020-09-10 11-02-02](https://user-images.githubusercontent.com/11939909/92706781-164aec00-f355-11ea-90e6-1d62246a16db.png)

jstastny-cz commented:
Issue#7
BC - Add to DRD does not seem to work.
1. Add Decision
2. DRD Actions -> Create
3. DRD Actions -> Add to new-diagram
4. Nothing happens

vpellegrino commented:
> Issue#7
> BC - Add to DRD does not seem to work.
> 
> 1. Add Decision
> 2. DRD Actions -> Create
> 3. DRD Actions -> Add to new-diagram
> 4. Nothing happens

@jstastny-cz is that issue present in the most recent version of source files? I believe that the service responsible for performing the DRD actions you described, has been committed only yesterday evening.

jstastny-cz commented:
> > Issue#7
> > BC - Add to DRD does not seem to work.
> > 
> > 1. Add Decision
> > 2. DRD Actions -> Create
> > 3. DRD Actions -> Add to new-diagram
> > 4. Nothing happens
> 
> @jstastny-cz is that issue present in the most recent version of source files? I believe that the service responsible for performing the DRD actions you described, has been committed ([e86c470](https://github.com/kiegroup/kie-wb-common/commit/e86c470ccc7c1b97962c6e50cea406b368e6fe6e)) only yesterday evening. Thanks for reporting

@vpellegrino I am working with war that @karreiro provided to me yesterday my afternoon, so timezone dependent - that might be why. I will retest with more recent build for sure. Thanks.

jstastny-cz commented:
Issue#8
- There's no clear indication of what DRD or DRG I am currently seeing (other than checking the decision components present on canvas).
- It would be nice if the currently selected node on canvas would be highlighted in the Decision Navigator too.
![Screenshot from 2020-09-10 12-29-10](https://user-images.githubusercontent.com/11939909/92717894-4e582c00-f361-11ea-9e9d-53f94cf176f5.png)

jstastny-cz commented:
Issue#9
- Decision Navigator enables navigation just using the DRD/DRG name, nothing happens when clicking decision components in the lists.
- In screenshot below, `a`, `new-diagram` and others do navigate to the respective items, but `Decision-1` or `Decision-2` do not. This might be considered together with Issue#8 - I would expect to navigate to correct DRD/DRG + highlight the respective Decision Component on canvas.
![Screenshot from 2020-09-10 12-51-02](https://user-images.githubusercontent.com/11939909/92720055-62515d00-f364-11ea-8b35-a3a6c8c02534.png)

karreiro commented:
Thank you for the new findings, @jstastny-cz.

The **Issue#6** is solved now (I've updated the artifacts in the PR description).

I couldn't reproduce the **Issue#7**, so I believe @vpellegrino is correct :-) (however, if we're missing something, please let us know).

Wrt **Issue#8**, those are excellent suggestions. Given the time frame, I believe we can follow UX team prototypes for now, but I'm going to hold these feedbacks as well. So, we can handle them on a future release when we iterate over this. I agree that having some indication when the DRG is opened will be nice.

I'm almost finishing the fix for the issue **Issue#4**, and right after, I'm going to check the **Issue#9**. As soon as I finish both, I will update the PR.

karreiro commented:
Jenkins do fdb.

jstastny-cz commented:
Issue#10
- unable to open previously valid files (project sent via email).

karreiro commented:
Jenkins do fdb.

karreiro commented:
Thank you, @jstastny-cz. I've just checked project files, and the model that does not open is a DMN 1.1 model. This PR temporarily removes DMN 1.1 support (I sent more details via email) :-)

jstastny-cz commented:
Testing with latest FDB.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorItemsProvider.java:
```diff
@@ -0,0 +1,114 @@
+
+        if (items.size() > 1) {
+            items.add(0, itemFactory.makeSeparator(DRG));
+            items.add(2, itemFactory.makeSeparator(DRDs));
```
I believe we already clarified this in the GChat. So, just reinforcing that each DMN model has a single DRG.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/factories/DecisionNavigatorItemFactory.java:
```diff
@@ -51,4 +68,73 @@ public DecisionNavigatorItem makeItem(final Node<View, Edge> node) {
+        }
+
+        return navigatorItemBuilder()
+                .withUUID(uuid)
```
This UUIDs are generated/managed by us, and at this context we don't need to process such checks :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/definition/model/TextAnnotation.java:
```diff
@@ -57,12 +58,18 @@
     protected boolean allowOnlyVisualChange;
 
+    /**
+     * Hold the {@link DMNDiagramElement} id for the {@link TextAnnotation} instance.
```
I believe that you're proposing to create this relationship by using a `DMNDiagramElement` object, instead of the diagram id. It's an interesting suggestion. Personally, I prefer to keep this kind of reference on text annotations / DRG elements as light as possible.

Given our timeframe and the stability of this PR, I believe we can keep it as is for now, and if the code evolve in a direction we need more details about the DMN diagram elements, we can refactor text annotations / DRG elements on this direction :-)

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/included/components/DecisionComponents.java:
```diff
@@ -57,17 +57,23 @@
 
+    private final DMNDiagramsSession dmnDiagramsSession;
+
+    public DecisionComponents() {
```
Some classes on kie-wb-common need this kind of empty "// CDI proxy" constructor implementation. I've added a comment to clarify. Thanks, @jstastny-cz.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-webapp-kogito-common/src/main/java/org/kie/workbench/common/dmn/webapp/kogito/common/client/included/DMNMarshallerImportsContentServiceKogitoImpl.java:
```diff
@@ -0,0 +1,101 @@
+    public Promise<PMMLDocumentMetadata> getPMMLDocumentMetadata(final String pmmlFilePath) {
+
+        if (StringUtils.isEmpty(pmmlFilePath)) {
+            return promises.reject("PMML file required to be marshalled is empty or null");
```
Fixed! Thanks!

karreiro commented:
Jenkins do fdb.

karreiro commented:
Hi, @jstastny-cz,

I've just updated the PR description with the most stable artifacts. They incorporate fixes for the Issue#4, the Issue#9, and the other fixes for issues I've found during manual tests (thank you, @vpellegrino @danielzhe).

Also, I've left inline answers for the comments.

Please, let me know if you need anything else :-)

jstastny-cz commented:
Jenkins do fdb.

jstastny-cz commented:
Issue#10
- Move node to a DRD
- navigate to the DRG
- delete the node
- Undo
- error appears and when confirmed, the DRG is broken (DRD looks fine)
![drd-undo](https://user-images.githubusercontent.com/11939909/93185473-bcc03400-f73d-11ea-9e73-d1276a14d277.gif)

jstastny-cz commented:
Issue#12
- Using Traffic Violation sample
- Move Fine and Violation nodes to new DRD
- run validation
- numerous validation errors flickers as notification, 2 data type related found in Alerts (first one is present also in original dmn file):
```
"Level","Text","File","Column","Line"
"WARNING","[_0C25EABB-852E-441A-97B2-1971933EC1EB] DMN: Columns: [Violation.Type] relate to FEEL string values which can be enumerated for the inputs; Gap analysis skipped. (DMN id: _0C25EABB-852E-441A-97B2-1971933EC1EB, DMN Validation, Decision Table Analysis, Gap Analysis) ","Traffic Violation.dmn","0","0"
"ERROR","[_75C69423-3FB6-4F20-BE52-464BCF0FFAD8] DMN: Unable to resolve type reference 'tFine' on node 'Fine' (DMN id: _75C69423-3FB6-4F20-BE52-464BCF0FFAD8, The listed type definition was not found) ","Traffic Violation.dmn","0","0"
"ERROR","[_F82F41FB-A8FB-40B5-A797-441ED2893F6E] DMN: Unable to resolve type reference 'tViolation' on node 'Violation' (DMN id: _F82F41FB-A8FB-40B5-A797-441ED2893F6E, The listed type definition was not found) ","Traffic Violation.dmn","0","0"
```
![type-validation](https://user-images.githubusercontent.com/11939909/93190251-6950e480-f743-11ea-8627-bcf71b7983b3.gif)

jstastny-cz commented:
Issue#13
- Pick a Decision Component
- drop it into Decision Navigator
- Following errors show up
IMO the UI should either handle this by adding the component into respective DRD (based on position in Decision Navigator) or allow pasting just over canvas, not in other locations.
![drag-drop](https://user-images.githubusercontent.com/11939909/93191296-a5387980-f744-11ea-98a3-772e780c57a8.gif)


```
"Level","Text","File","Column","Line"
"ERROR","Command execution failed.Reason/s: [1] The 'org.kie.workbench.common.dmn.api.DMNDefinitionSet' cannot accepts the containment for the candidate roles '[org.kie.workbench.common.dmn.api.definition.model.InputData, input-data]'","Traffic Violation.dmn","0","0"
```

jstastny-cz commented:
Issue#14
- create DMN with a single decision
- don't do any other configuration and save
- close
- open
- click the decision, click Edit
- Error notification appears
![dmn-save-open](https://user-images.githubusercontent.com/11939909/93228459-ca91ab80-f775-11ea-9bf0-c93e28360848.gif)

jstastny-cz commented:
Issue#15
the user wants to simplify the DRG diagram
- moves components to DRD
- wants to truncate DRG graph of the details that are being hidden
- the components are removed globally.

![move-to-drd-and-remove](https://user-images.githubusercontent.com/11939909/93231585-2d387680-f779-11ea-9bbc-a0ee74fe1f7a.gif)

jstastny-cz commented:
Issue#16
- create DMN with Decision
- define Decision Table
- first column set as number
- save asset, close editor
- reopen asset
- go to the decision table, select the input with type number
- open properties panel, set constraint
- move elsewhere
- if moved back, the constraint value is gone
![dmn-constraint](https://user-images.githubusercontent.com/11939909/93230733-48ef4d00-f778-11ea-8184-29981f9cde9f.gif)

jstastny-cz commented:
Issue#17
- In DRG have a decision component (Decision)
- in a DRD connect an input node to the Decision (which has been added from DRG)
- Save
- open again
- go to DRD -> the connection is missing
![relation-not-saved](https://user-images.githubusercontent.com/11939909/93234890-290e5800-f77d-11ea-90ef-314a4185856b.gif)

vpellegrino commented:
> Issue#16
> 
> * create DMN with Decision
> * define Decision Table
> * first column set as number
> * save asset, close editor
> * reopen asset
> * go to the decision table, select the input with type number
> * open properties panel, set constraint
> * move elsewhere
> * if moved back, the constraint value is gone
>   ![dmn-constraint](https://user-images.githubusercontent.com/11939909/93230733-48ef4d00-f778-11ea-8184-29981f9cde9f.gif)

Hi @jstastny-cz 
I just tried to reproduce that issue, but in my case the flow is interrupted when I click on `save` button:
![image](https://user-images.githubusercontent.com/22591802/93314312-e3e43780-f809-11ea-9719-87b42c6470e9.png)
Can you please verify if with the latest artifact the behaviour is the same? which webapp are you using locally? 

Thanks

jstastny-cz commented:
@vpellegrino I am using the v4 WAR from the description.

jstastny-cz commented:
Jenkins do fdb.

jstastny-cz commented:
jenkins retest this

jstastny-cz commented:
FDB reporting checkstyle failure in kie-wb-common-stunner-client-common - please check the changes in this PR in given module.

vpellegrino commented:
> FDB reporting checkstyle failure in kie-wb-common-stunner-client-common - please check the changes in this PR in given module.

Just to have a cleaner overview,
I was able to find in the logs the checkstyle failure:

```
[2020-09-16T09:53:18.671Z] [INFO] There is 1 error reported by Checkstyle 8.19 with /home/jenkins/workspace/KIE/master/fullDownstream/kie-wb-common-master.downstream/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/target/checkstyle-rules.xml ruleset.
[2020-09-16T09:53:18.671Z] [ERROR] src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/UndoSessionCommand.java:[1] (header) RegexpHeader: Line does not match expected header line of '^\/\*$'.
```

karreiro commented:
@vpellegrino The error you've link in the screenshot is a validation error. It's just alerting you about the input column, but it does not block the save operation.

jstastny-cz commented:
> Moreover, do you feel that reported issue is related to multiple DRDs change? Have you tried the same on master branch?

@vpellegrino yes, I can't see this in master. I suspect the marshaller changes, but I am just guessing.

karreiro commented:
Jenkins do fdb.

karreiro commented:
Thank you for the review comments, @jstastny-cz.

Regarding each pending issue we have:

- **Issue#11** - I've raise [a JIRA](https://issues.redhat.com/browse/KOGITO-3364) for this. @danielzhe will solve that on a separate PR

- **Issue#12** - Fixed.

- **Issue#13** - This issue wasn't introduced by this PR. I've create [this JIRA](https://issues.redhat.com/browse/DROOLS-5646) to address it.

- **Issue#14** - Fixed.

- **Issue#15** - Actually, when users remove a DRG element from the DRG diagram, the expected behaviour is to remove all references of that element across all the diagrams.

- **Issue#16** - @vpellegrino is working on a fix for this. I've created [this JIRA](https://issues.redhat.com/browse/KOGITO-3366) and the fix will be submitted on a separate PR

- **Issue#17** - Fixed.

I've just updated the PR description with the most stable artifacts.

Thanks, @jstastny-cz.

karreiro commented:
Jenkins do fdb.

jstastny-cz commented:
Issue#18
- unzip the git repo
- import into space using file:/// address
- go to `urgent` dmn model
- validate
I am getting sort of contradictory validation messages for the asset:
```
"Level","Text","File","Column","Line"
"WARNING","[_AAFBE7A4-9B4C-4A42-9CF3-06B1B8BA81B2] DMN: Skipped Decision Table Analysis of table 'Suitable Rooms' because: Unable to locate typeRef {https://kiegroup.org/dmn/_46BDFA04-740C-425F-BCB9-853271AAFB22}UNDEFINED to determine domain. (DMN id: _AAFBE7A4-9B4C-4A42-9CF3-06B1B8BA81B2, DMN Validation, Decision Table Analysis) ","urgent.dmn","0","0"
"WARNING","[_06222075-C3A0-43AF-A534-8345D070BB71] DMN: Decision table with single output on node 'Suitable Rooms' should not have an output type reference (DMN id: _06222075-C3A0-43AF-A534-8345D070BB71, The referenced element should not have a typeref set) ","urgent.dmn","0","0"
"ERROR","[_06222075-C3A0-43AF-A534-8345D070BB71] DMN: Unable to resolve type reference 'UNDEFINED' on node 'Suitable Rooms' (DMN id: _06222075-C3A0-43AF-A534-8345D070BB71, The listed type definition was not found) ","urgent.dmn","0","0"
```
Please check the files, I wasn't to able to find any occurence of UNDEFINED in the dmn file itself.
[hospital.zip](https://github.com/kiegroup/kie-wb-common/files/5236814/hospital.zip)

jstastny-cz commented:
ad Issue#15
we can't remove the component globally when removing from DRG IMO, consider following:
- define Decision in DRD
- go to DRG and reuse the top-level decision from DRD (Notice that the component is now owned by DRG)
- Realize you don't need it in DRG
- Remove from DRG
- it gets removed globally

IMO the solution for now might be a dialog where user can a) remove globally, b) remove from DRG only, c) cancel
WDYT?

![remove-drg](https://user-images.githubusercontent.com/11939909/93464117-f1bcba00-f8e8-11ea-9537-a069841c2357.gif)

jstastny-cz commented:
Marking resolved issues by editing my comments as (Fixed). 
We need to create jiras for:
- Issue#2,
- Issue#3,
- Issue#5,
- Issue#8,
- Issue#10

danielzhe reviewed (approved):
Looks good to me. Just need the fix for Issue#18 that @karreiro is working on.

karreiro commented:
@jstastny-cz,

Regarding **Issue#2**, **Issue#3**, **Issue#8** - I've created this UX JIRA to follow up them https://issues.redhat.com/browse/DROOLS-5651.

The **Issue#5** will by addressed by this: https://issues.redhat.com/browse/KOGITO-3306.

Also, I've created the https://issues.redhat.com/browse/KOGITO-3379 to address **Issue#15**. But, before implementing it, I will get some UX feedback and confirm if it's the right direction.

Regarding the **Issue#10**, we already have a JIRA for that, it's the BAPL-1489.

Wrt **Issue#18**. It's important to mention **it wasn't introduced by this PR**, it was a bug in the Kogito client side marshaller. Still, I've fixed it in the latest commit and updated the PR description with the most stable artifacts.

I agree with @danielzhe. This ensemble looks good enough to me too, +1 for merging it and solving other enhancements/issues on separated PRs.

karreiro commented:
Jenkins do fdb.

vpellegrino reviewed (approved):
LGTM!

jstastny-cz reviewed (approved):
@karreiro  Thanks for all the changes, good to merge.
